conwayste/src/client.rs

```diff
@@ -413,41 +430,105 @@ impl EventHandler for MainState {
-            Stage::Run => {
-                // while this works at limiting the FPS, its a bit glitchy for input events... that probably should be
-                // rewritten as it was hacked together originally
-               // while timer::check_update_time(ctx, FPS)
+            Screen::Run => {
+                // TODO Disable FSP limit until we decide if we need it
+                // while timer::check_update_time(ctx, FPS) {
+                let mut textfield_under_focus = false;
+                if let Some(tf) = self.ui_manager.textfield_from_id(Screen::Run, WidgetID::InGamePane1ChatboxTextField) {
+                    match tf.state {
+                        Some(TextInputState::TextInputComplete) =>  {
+                            textfield_under_focus = false;
+                            self.handle_user_chat_complete(ctx)?;
+                        }
+                        Some(TextInputState::EnteringText) => {
+                            textfield_under_focus = true;
+                            tf.update(ctx)?;
+                        },
+                        None => {
+                            textfield_under_focus = false;
+                            tf.update(ctx)?;
+                        },
+                    }
+                }
+
+                if textfield_under_focus {
+                    self.process_text_field_inputs(ctx);
+                } else {
+                    self.process_running_inputs();
+                }
 
-                self.process_running_inputs();
+                if self.inputs.mouse_info.mousebutton == MouseButton::Left {
+                    let (x,y) = self.inputs.mouse_info.position;
+                    let mouse_pos = Point2::new(x as f32, y as f32);
+
+                    fn flip_cell(ms: &mut MainState, cell: Cell) {
+                        // Make dead cells alive or alive cells dead
+                        let result = ms.uni.toggle(cell.col, cell.row, CURRENT_PLAYER_ID);
+                        ms.drag_draw = match result {
+                            Ok(state) => Some(state),
+                            Err(_)    => None,
+                        };
+                    }
+
+                    match self.inputs.mouse_info.action {
+                        Some(MouseAction::Click) => {
+                            if let Some(cell) = self.viewport.get_cell(mouse_pos) {
+                                flip_cell(self, cell)
+                            }
+                        }
+                        Some(MouseAction::Drag) => {
+                            if let Some(cell) = self.viewport.get_cell(mouse_pos) {
+                                // Only make dead cells alive
+                                if let Some(cell_state) = self.drag_draw {
+                                    self.uni.set(cell.col, cell.row, cell_state, CURRENT_PLAYER_ID);
+                                } else {
+                                    flip_cell(self, cell)
```

Comment:
I'm confused about what's going on here.

Reply:
There's a bug in here related to how drag-n-draw clears the final cell, as you pointed out offline.

But the gist of it is that if we're dragging the mouse around, it will set the cell state to alive. If it's just a mouse click (the `else` case, not dragging), then we toggle the cell's state.